drools-wb-screens/drools-wb-guided-dtable-editor/drools-wb-guided-dtable-editor-client/src/main/java/org/drools/workbench/screens/guided/dtable/client/wizard/table/pages/FactPatternsPageViewImpl.java

```diff
@@ -149,275 +146,272 @@ public void setup() {
+                chosenPatterns.add(index + 1,
+                                   p);
+                setChosenPatterns(chosenPatterns);
+                presenter.setConditionPatterns(chosenPatterns);
             }
-
-        } );
+        });
     }
 
     @Override
-    public void setAvailableFactTypes( final List<String> types ) {
-        availableTypesWidget.setRowCount( types.size(),
-                                          true );
-        availableTypesWidget.setRowData( types );
+    public void setAvailableFactTypes(final List<String> types) {
+        availableTypesWidget.setRowCount(types.size(),
+                                         true);
+        availableTypesWidget.setRowData(types);
     }
 
     @Override
-    public void setChosenPatterns( final List<Pattern52> types ) {
+    public void setChosenPatterns(final List<Pattern52> types) {
         chosenPatterns = types;
-        chosenPatternWidget.setRowCount( types.size(),
-                                         true );
-        chosenPatternWidget.setRowData( types );
+        chosenPatternWidget.setRowCount(types.size(),
+                                        true);
+        chosenPatternWidget.setRowData(types);
         enableMoveUpButton();
         enableMoveDownButton();
         presenter.stateChanged();
     }
 
     @Override
-    public void init( final FactPatternsPageView.Presenter presenter ) {
+    public void init(final FactPatternsPageView.Presenter presenter) {
         this.presenter = presenter;
     }
 
     @Override
-    public void setArePatternBindingsUnique( final boolean arePatternBindingsUnique ) {
+    public void setArePatternBindingsUnique(final boolean arePatternBindingsUnique) {
         chosenPatternWidget.redraw();
     }
 
     @UiHandler(value = "btnAdd")
-    public void btnAddClick( final ClickEvent event ) {
-        for ( String type : availableTypesSelections ) {
+    public void btnAddClick(final ClickEvent event) {
+        for (String type : availableTypesSelections) {
             final Pattern52 pattern = new Pattern52();
-            pattern.setFactType( type );
-            chosenPatterns.add( pattern );
+            pattern.setFactType(type);
+            chosenPatterns.add(pattern);
         }
-        setChosenPatterns( chosenPatterns );
-        presenter.setConditionPatterns( chosenPatterns );
+        setChosenPatterns(chosenPatterns);
+        presenter.setConditionPatterns(chosenPatterns);
         presenter.stateChanged();
     }
 
     @UiHandler(value = "btnRemove")
-    public void btnRemoveClick( final ClickEvent event ) {
+    public void btnRemoveClick(final ClickEvent event) {
         boolean allPatternsRemoved = true;
-        for ( Pattern52 pattern : chosenPatternSelections ) {
-            if ( !validator.canPatternBeRemoved( pattern ) ) {
+        for (final Pattern52 pattern : new ArrayList<Pattern52>(chosenPatternSelections)) {
```

Comment:
@karreiro could you please tell me more about this change? I mean I am not sure why is needed to wrap `chosenPatternSelections` into new array list.